guides/release/typescript/core-concepts/invokables.md

```diff
@@ -0,0 +1,733 @@
+    let { onEnter, onExit, options } = named;
+    this.manager.observe(el, options, { onEnter, onExit });
+  }
+}
+```
+
+Notice that we can just skip the positional arguments entirely in this case, and give them a name like `_` to indicate we are doing nothing with it. If we had positional arguments, we would supply them like normal.
+
+## Advanced signature techniques
+
+We can also define signatures in more complicated ways using more advanced TypeScript features.
+Nearly anything you can do with a “regular” TypeScript function or class, you can also do with signatures for Glimmer invokables.
+We can make a component accept a [generic][generic] type, or use [union][union] types.
+With these tools at our disposal, we can even define our signatures to [make illegal states un-representable][illegal].
+
+To see this in practice, consider a list component which yields back out instances of the same type it provides, and provides the appropriate element target based on a `type` argument.
+Yielding back out the same type passed in will use generics, and providing an appropriate element target for `...attributes` can use a union type.
+
+Here is how that might look, using a class-backed component rather than a template-only component, since the only places TypeScript allows us to name new generic types are on functions and classes:
+
+```typescript
+import Component from '@glimmer/component';
+
+interface OrderedList<T> {
+  Args: {
+    items: Array<T>;
+    type: 'ordered';
+  };
+  Blocks: {
+    default: [item: T];
+  };
+  Element: HTMLOListElement;
+}
+
+interface UnorderedList<T> {
+  Args: {
+    items: Array<T>;
+    type: 'unordered';
+  };
+  Blocks: {
+    default: [item: T];
+  };
+  Element: HTMLUListElement;
+}
+
+type ListSignature<T> = OrderedList<T> | UnorderedList<T>;
+
+export default class List<T> extends Component<ListSignature<T>> {
+  <template>
+    {{#if (isOrdered @type)}}
+      <ol ...attributes>
+        {{#each @items as |item|}}
+          <li>{{yield item}}</li>
+        {{/each}}
+      </ol>
+    {{else}}
+      <ul ...attributes>
+        {{#each @items as |item|}}
+          <li>{{yield item}}</li>
+        {{/each}}
+      </ul>
+    {{/if}}
+  </template>
+}
+
+function isOrdered(type: 'ordered' | 'unordered'): type is 'ordered' {
+  return type === 'ordered';
+}
+```
+
+If you are using Glint, when this component is invoked, the `@type` argument will determine what kinds of modifiers are legal to apply to it. For example, if you defined a modifier `reverse` which required an `HTMLOListElement`, this invocation would be rejected:
+
+```handlebars
+<List @items={{array 1 2 3}} @type='unordered' {{reverse}} as |item|>
+  The item is
+  {{item}}.
+</List>
+```
+
+The same approach with generics works for class-based helpers and class-based modifiers.
+Function-based helpers as and modifiers can also use generics, but by using them on the function definition rather than via a signature.
```

Comment:
```suggestion
Function-based helpers and modifiers can also use generics, but by using them on the function definition rather than via a signature.
```